remove `HierarchyNotRespected` validation error

cdisselkoen opened the pull request:
## Description of changes

As suggested in #638, removes the `HierarchyNotRespected` validation error (well, except from the public error enum, where we leave it to avoid a breaking change, but adjust documentation to indicate it is never returned).  This simplifies the typechecker code considerably.  

I'm a little unclear from the discussion on #638 whether the Lean model needs any corresponding changes or not.

## Issue #, if available

Fixes #638 

## Checklist for requesting a review

The change in this PR is (choose one, and delete the other options):

- [x] A bug fix or other functionality change requiring a patch to `cedar-policy`.

I confirm that this PR (choose one, and delete the other options):

- [x] Updates the "Unreleased" section of the CHANGELOG with a description of my change (required for major/minor version bumps).

I confirm that [`cedar-spec`](https://github.com/cedar-policy/cedar-spec) (choose one, and delete the other options):

- Not sure about changes to the Lean model, see above

I confirm that [`docs.cedarpolicy.com`](https://docs.cedarpolicy.com/) (choose one, and delete the other options):

- [x] Does not require updates because my change does not impact the Cedar language specification.

john-h-kastner-aws reviewed (approved):
I happy to see this changes,  though I'm not 100% sure it will pass DRT. It might be returning `Bool` when the lean impl gives a more precise `False` type. Fixing any discrepancy there should also be a backwards compatible change, but we should take a close look at what both are doing with `in`.

If it does pass, then we _might_ be able to use `AgreeOnAll` here if this is the only error the Rust reports but Lean doesn't.

https://github.com/cedar-policy/cedar-spec/blob/main/cedar-drt/src/lean_impl.rs#L516

cdisselkoen commented on cedar-policy-validator/src/diagnostics/validation_errors.rs:
```diff
@@ -460,28 +460,19 @@ impl Diagnostic for FunctionArgumentValidation {
-#[error("for policy `{policy_id}`, operands to `in` do not respect the entity hierarchy")]
+#[error("Internal invariant violated: `HierarchyNotRespected` error should never occur. Please file an issue")]
 pub struct HierarchyNotRespected {
     /// Source location
```
`wrap_core_error!` in `api/err/validation_errors.rs` needs them.  Taking a couple steps back, since the variant still exists in the public `ValidationError`, we need to support `ValidationError::policy_id()` on it.  And you could say we could return a fictitious policy id or something, but if this error actually ever does get thrown in violation of the internal invariant (due to some future refactor), we'll be glad to have the policy id and source location

cdisselkoen commented on cedar-policy-validator/src/typecheck.rs:
```diff
@@ -1992,85 +1990,21 @@ impl<'a> Typechecker<'a> {
-                    } else if matches!(type_of_in.data(), Some(Type::False)) {
+                    if self.mode.is_strict() && matches!(type_of_in.data(), Some(Type::False)) {
                         TypecheckAnswer::success(
                             ExprBuilder::with_data(Some(Type::False)).val(false),
```
I'll do that in a separate PR since this one is a little delicate as-is and the change you suggest is technically orthogonal

cdisselkoen replied:
I think that change was done in #1462?

cdisselkoen replied:
Updated this PR, now simplified further as you say.  (we just return `success(type_of_in)` in all cases)

cdisselkoen commented:
> It might be returning `Bool` when the lean impl gives a more precise `False` type.

I looked at this Rust code some, and I think/hope that the logic I removed was already performed in `type_of_var_in_entity_literals()` and friends.  The logic I removed only ever results in `HierarchyNotRespected` or `success(type_of_in)`; it never returns `False` when `type_of_in` wasn't already `False`.  So removing this logic shouldn't stop us from returning `False` in any situation where we were already returning `False` before this PR.

john-h-kastner-aws commented:
> > It might be returning `Bool` when the lean impl gives a more precise `False` type.
> 
> I looked at this Rust code some, and I think/hope that the logic I removed was already performed in `type_of_var_in_entity_literals()` and friends. The logic I removed only ever results in `HierarchyNotRespected` or `success(type_of_in)`; it never returns `False` when `type_of_in` wasn't already `False`. So removing this logic shouldn't stop us from returning `False` in any situation where we were already returning `False` before this PR.

IIUC, the DRT currently passes if Rust returns an errors when lean doesn't. So, if rust has stopped returning errors here, there could be a precision difference in the non-error response.

john-h-kastner-aws reviewed (approved):
This looks good to me. The concerns I mention will be fully addressed just by running the diff tests with this change merged, so we just need another review to look this over.

cdisselkoen commented:
Corpus tests are failing in CI, probably because there are edge cases where this PR changes the validation behavior from fail -> pass (as noted in the changelog entry).  Adjusting the corpus test expectation will have to be a different PR against cedar-integration-tests

cdisselkoen commented:
note on the status of this: have been waiting for cedar-policy/cedar-integration-tests#14 to be merged first, to avoid conflicts in updating the corpus tests

cdisselkoen commented:
Ready for re-review. Corpus test failure is expected, see comments above; we will need to regenerate corpus tests because of this PR, which will be a separate PR to cedar-integration-tests. Also, DRT failure is because as of this writing we have merged #1535 but not cedar-policy/cedar-spec#587.
